### ISSUE TITLE
Performance Issue with Feature Constructor

### ISSUE THREAD

**Issue by mw25** [biolab/orange3#5911]
**What's wrong?**

It's not about a bug, rather a performance issue.

I have already noticed that Feature Constructor takes a little longer with more data.
But now I want to use the construct_variables function in my own widget as well, so I took a closer look:

I analyzed my minimal example (see below) with py-spy and saw that the function `__call__` of FeatureFunc is the bottleneck. Especially collecting the data line by line, it seems to me.
I could imagine saving a lot of time by first reading out the data column-by-column (and then performing the operation row-by-row again, if necessary). 
Of course, I don't know what impact that would have on the further logic and design of the functions in the widget....

Adding two 100,000 long variables took a little less than 2 seconds in my case.

What do you think? Could you imagine to improve the performance?

**How can we reproduce the problem?** 

```
    def test_performance(self):
        from codetiming import Timer
        timer = Timer()

        var_names = ['Var' + str(i) for i in range(100)]
        variables = [ContinuousVariable(name=var_name) for var_name in var_names]
        domain = Domain(variables)
        test_data = np.random.rand(100_000, len(variables))
        table = Table.from_numpy(domain, test_data)

        expr = 'Var0 + Var1'
        desc = ContinuousDescriptor(expr, expr, number_of_decimals=None)
        constr_vars = construct_variables([desc], table)
        new_domain = Domain(table.domain.attributes + tuple(constr_vars))

        timer.start()
        new_table = table.transform(new_domain)
        timer.stop()
```

Console Output: 
`Elapsed time: 1.8722 seconds`

The result of py-spy:

![image](https://user-images.githubusercontent.com/67997227/159276343-9ed20bba-c302-447e-8725-edca504714d0.png)

**What's your environment?**

- Operating system: Windows 10
- Orange version: '3.32.0.dev0+2b857a3'
- How you installed Orange: git clone

**Comment by mw25** [comment]
Ahhh, thank you very much! Reading the documentation always helps...

In my case I need to apply the calculation of a variable created in Feature Constructor to new data. Unfortunately this doesn't work with whole columns like in my or your simple example. 
So I have passed to `compute_value` a function which will read out the necessary source data column by column and then compute the expression from Feature Constructor value by value.
This still gives me an improvement of factor 20, which is enough for now.

**Comment by janezd** [comment]
As far as I see, the function is compiled once.

The interesting part is here: https://github.com/biolab/orange3/blob/master/Orange/widgets/data/owfeatureconstructor.py#L1166. If this could be avoided, the function could be computed on vectors.

This would be possible, in principle, by simply computing `args` (a few lines later) on table instead of on instance (e.g. preferably using `get_column_view`, and being careful with sparse data and strings). However, not all user-defined expressions are computable in such fashion. Perhaps we could try with vectors first, and if this fails, fall back to per-instance?

**Comment by janezd** [comment]
Resolved in #5975.
